Sms verification

86667 opened the pull request.

86667 committed:
feat: Prelude and SmsVerification services

86667 committed:
feat: Db and integration into smsVerificationService

86667 committed:
test: e2e tests

86667 committed:
feat: Homeserver admin api

86667 committed:
feat: wire everything together

86667 committed:
feat: max 10 verifications per phone number

86667 committed:
fix: grab ip address from Request

86667 committed:
chore: refactoring and cleaning up

86667 committed:
feat: tightening logic and state transitions"

86667 committed:
test: http tests

86667 committed:
fix: Some more types cleanup and refactoring

SeverinAlexB commented on .env.example:
```diff
@@ -5,3 +5,7 @@ DATABASE_URL=postgres://localhost:5432/pubky_homegate
 # Prelude.so API key
 HG_PRELUDE_API_KEY=
+# Homeserver admin API URL
+HG_HOMESERVER_API_URL=http://127.0.0.1:6288
```
```suggestion
# Homeserver admin API URL
HG_HOMESERVER_ADMIN_API_URL=http://127.0.0.1:6288
```

SeverinAlexB commented on src/infrastructure/config.rs:
```diff
@@ -10,6 +11,21 @@ pub struct EnvConfig {
+    pub homeserver_admin_password: String,
+    pub homeserver_pubky: String,
+    #[serde(default = "default_max_verified_sessions")]
+    pub max_verified_sessions: u32,
```
Missing in .env.example

SeverinAlexB commented on .env.example:
```diff
@@ -5,3 +5,7 @@ DATABASE_URL=postgres://localhost:5432/pubky_homegate
 HG_HTTP_LISTEN_SOCKET=127.0.0.1:5000
 # Prelude.so API key
 HG_PRELUDE_API_KEY=
+# Homeserver admin API URL
```
Requires a better description. What is this exactly? Where to take the value from?

86667 replied:
Ive updated a little but not added any more of a description. What other info would you like here?

SeverinAlexB replied:
I think this is good.

SeverinAlexB commented on src/app_state.rs:
```diff
@@ -0,0 +1,68 @@
+};
+
+#[derive(Clone, Debug)]
+pub struct AppState<T, S>
```
Let's keep the naming uniform over multiple projects.
- `AppState` -> Axum webserver state
- `AppContext` -> Application context to bootstrap the system.

SeverinAlexB commented on src/e2e/mod.rs:
```diff
@@ -0,0 +1,92 @@
+    pub client: Client,
+}
+
+impl TestServer {
```
Why not use https://docs.rs/axum-test/latest/axum_test/?

86667 replied:
Good point, updated.

SeverinAlexB commented on src/http_server/ip_extraction.rs:
```diff
@@ -0,0 +1,122 @@
+use std::net::SocketAddr;
+
+/// Extract client IP address from request, checking proxy headers first
+pub fn extract_client_ip(addr: SocketAddr, headers: &HeaderMap) -> String {
```
Why return a String instead of an IpAddr? This would also make sure that the header is an actual IP and not gibberish.

Reminds me of https://refactoring.guru/smells/primitive-obsession

SeverinAlexB commented on src/infrastructure/http/extractors.rs:
```diff
@@ -0,0 +1,122 @@
+                return trimmed.to_string();
+            }
+        }
+    }
```
This code was actually very readable. Why did you rewrite it?
https://github.com/pubky/pkarr/pull/185/files#diff-7da92a66fcd6ee529cafe90e1fa1e4a7311a3ade4ed3ed931cf85da8f462c07cR45

SeverinAlexB reviewed (changes requested):
I haven't deeply reviewed it yet. From what I see, the code structure doesn't follow the modules approach. Instead it splits up the code in technical layers again.

Instead of having one sms module including the prelude api, the http router and so on, you split up the code in technicalities
- external apis
- e2e
- client_server/routes
- sms_verification

I will do a second round later.

SeverinAlexB commented:
Also, might be worth considering [wiremock](https://crates.io/crates/wiremock) for mocking http requests instead of using traits. I haven't used it yet so I am curious what you think about it.

There are multiple libs like this https://github.com/lipanski/mockito

86667 committed:
fix: Small changes for PR review

86667 committed:
feat: module-based code structure

SeverinAlexB commented on src/sms_verification/app_state.rs:
```diff
@@ -0,0 +1,93 @@
+pub struct AppState<T, S>
+where
+    T: SmsVerificationProviderApi + Clone + 'static,
+    S: HomeserverAdminApiTrait + Clone + 'static,
+{
+    pub db: SqlDb,
+    pub sms_verification: SmsVerificationService<T, S>,
+}
```
Instead of having a shared AppState between all the module (SMS and so on), each module could have it's own router and own app state. This way, you can keep the app state logic in their specific modules.

See this quick AI example
https://grok.com/share/bGVnYWN5LWNvcHk_2d0cdfdb-4276-49f6-8bb4-1e153a7c1fcb

SeverinAlexB commented on src/infrastructure/http/extractors.rs:
```diff
@@ -0,0 +1,122 @@
+use std::net::{IpAddr, SocketAddr};
+
+/// Extract client IP address from request, checking proxy headers first
+pub fn extract_client_ip(addr: SocketAddr, headers: &HeaderMap) -> IpAddr {
```
An extractor is a very specific thing in axum.
This method is not an extractor. I recommend actually using an extractor here to provide the client IP to all paths.

For the sake of training, please read up on [extractors](https://docs.rs/axum/latest/axum/extract/index.html#defining-custom-extractors) and implement it here.

For example, the extractor could be used like this then:

```rust
async fn send_code_handler(ExtractRequestOrigin(ip_addr): ExtractRequestOrigin) {
    // ...
}
```

86667 replied:
Good suggestion

SeverinAlexB commented on src/sms_verification/app_state.rs:
```diff
@@ -0,0 +1,93 @@
+pub struct AppState<T, S>
+where
+    T: SmsVerificationProviderApi + Clone + 'static,
+    S: HomeserverAdminApiTrait + Clone + 'static,
```
Strongly suggest using mocks here. Traits make the code very ugly and hard to read.
https://github.com/pubky/homegate/pull/2#issuecomment-3612022559

SeverinAlexB commented on src/infrastructure/http/extractors.rs:
```diff
@@ -0,0 +1,122 @@
+                return parsed_ip;
+            }
+        }
+    }
```
Comments like this are always a red flag. This problem shows up in multiple parts of the code.
Always remember:

> The best comment is a good name for a method or class.

https://refactoring.guru/smells/comments

SeverinAlexB commented on src/sms_verification/service.rs:
```diff
@@ -0,0 +1,219 @@
+    /// Validates that a phone number is in E.164 format
+    fn validate_phone_number(phone_number: &str) -> Result<(), SmsVerificationError> {
+        // E.164 format: starts with +, followed by 1-15 digits
+        let e164_regex = Regex::new(r"^\+[1-9]\d{1,14}$").unwrap();
```
Never use an unwrap in production code. If it is always valid, use `.expect(reason)` and write down exactly why this doesn't panic. This is a general rule.

SeverinAlexB replied:
In test code `unwrap()` is obviously no issue

SeverinAlexB reviewed (changes requested):
It's getting better. Let's keep improving the code structure. I will do an in details review as soon as this is nailed down.

86667 committed:
fix: rm regex unwrap and replace extract_ip()

86667 committed:
feat: user Axum extractor pattern to find request's origin

86667 committed:
feat: Split AppState per-module

86667 committed:
test: use wiremock. Remove traits

SeverinAlexB commented on src/shared/homeserver_admin_api.rs:
```diff
@@ -0,0 +1,64 @@
+use crate::EnvConfig;
```
No need for the `homeserver` folder here no? It is a single file anyway so I can live in `shared` directly.

SeverinAlexB commented on src/shared/homeserver/homeserver_admin_api.rs:
```diff
@@ -0,0 +1,64 @@
+            .await?;
+
+        let status = response.status();
+        if !status.is_success() {
```
NIT: I don't know if you know but response has a convenient `.error_for_status()` method which already provides a error struct. Not sure if it accomplished the same as you want here.

SeverinAlexB commented on src/shared/homeserver/homeserver_admin_api.rs:
```diff
@@ -0,0 +1,64 @@
+            return Err(HomeserverAdminApiError::ApiError {
+                status: status.as_u16(),
+                message: error_body,
+            });
```
Can you add the url here too? This will help associated the error with a specific request. I think the reqwest error struct would already include that.

SeverinAlexB commented on src/sms_verification/app_state.rs:
```diff
@@ -0,0 +1,52 @@
+        // Connect to database (migrations run automatically)
+        let db = SqlDb::connect(&config.database_url)
+            .await
+            .map_err(SmsVerificationRepositoryError::Database)?;
```
The goal is to have one shared database connection for the whole homegate app. This is not an issue currently but as soon as we have the lightning module too, we should not create a second connection.

SeverinAlexB commented on src/sms_verification/service.rs:
```diff
@@ -0,0 +1,177 @@
+                phone_number.to_string(),
+            ))
+        }
+    }
```
Phone number should be a struct.
https://refactoring.guru/smells/primitive-obsession

This way, `Json(request): Json<CreateVerificationRequest>` would also validate the phone number upon receving the request and return `400 bad request` as expected.

```rust 
async fn send_code_handler(
    State(state): State<AppState>,
    RequestOrigin(ip_address): RequestOrigin,
    Json(request): Json<CreateVerificationRequest>,
) -> Result<Json<CreateVerificationResponse>, SmsVerificationError> {
```

SeverinAlexB commented on src/sms_verification/service.rs:
```diff
@@ -0,0 +1,177 @@
+            return Err(SmsVerificationError::TooManyVerifiedSessions);
+        }
+        Ok(())
+    }
```
We will need a better verification logic here. Please ask @BitcoinErrorLog  what his thoughts are. My proposal:
- General IP rate limiting including LN payments and SMS: Max 5 signups per year. Needs an allowlist for conference WIFI ips though.
- SMS (rate limit per Phone): Max 2 signups per week, 4 per year

86667 replied:
Im not sure about IP limiting: 
- We're bound to at some point forget completely to add a conference IP or configure only one of several ips in a large venue, or something
- Cafes, bars etc that we dont know about could have a bunch of users onboarding at the same time in the not so distant future 
- Its easy and cheap to bypass if a user really wants to, anyway

Since LN verification is real money that may be a sufficient deterrent.

Sms is free for the user so we should limit that. 2 per week, 4 per year seems reasonable

86667 replied:
@BitcoinErrorLog's response - https://github.com/pubky/homegate/pull/2#issuecomment-3638683920

I'll implement 2 per week 4 per year in this PR.

SeverinAlexB replied:
Cool. But please still add some IP rate limiting so nobody can abuse the system. Two critical endpoints that should have some limits:
- Send SMS code: We pay for each sms. We don't want somebody with a single IP to come with a million phone numbers and hit us with a huge bill.
- Lightning invoice creation: Phoenixd is great but I am sure it will struggle with a million invoices. So same here: Let's add a reasonable rate limit so one IP address can't just create a million invoices.

No idea what is reasonable here. 50 SMS verifications per hour per IP? 50 Lightning invoices per hour per IP?

Does this make sense?

86667 replied:
Yes makes sense :+1: 

This sounds more like infrastructure level rate limiting rather than business logic limiting though. This would be better done at the load balancer? Or do we typically do this in application code for other services too?

SeverinAlexB replied:
I think this can be done with nginx. @SpontaneousOverthrow you can do that, can't you?

SpontaneousOverthrow replied:
yes, let's do that on balancer lvl, for doing that I need paths and limits that you want to apply to that paths

SeverinAlexB commented on src/sms_verification/prelude_api/client.rs:
```diff
@@ -0,0 +1,209 @@
+use crate::EnvConfig;
```
Unnecessary `prelude_api` folder. This could just be a file.

SeverinAlexB commented on src/sms_verification/http.rs:
```diff
@@ -0,0 +1,128 @@
+                tracing::error!(status = status, message = %message, "SMS provider API error");
+                (
+                    StatusCode::BAD_GATEWAY,
+                    "external_service_error",
+                    format!("SMS provider error ({}): {}", status, message),
+                )
+            }
```
Arent' there the same error?

They should be Internal Server Error. Bad Gateway means something else.

SeverinAlexB commented on src/sms_verification/prelude_api/client.rs:
```diff
@@ -0,0 +1,209 @@
+            .await
+            .map_err(|e| {
+                SmsVerificationError::InvalidResponse(format!("Failed to parse response: {}", e))
+            })?;
```
Please log this error with the response, otherwise we will never see it in the logs.

SeverinAlexB commented on src/sms_verification/repository.rs:
```diff
@@ -0,0 +1,208 @@
+    DatabaseError(#[from] DbError),
+
+    #[error("Database error: {0}")]
+    Database(#[from] sqlx::Error),
```
Identical errors?

SeverinAlexB commented on src/sms_verification/repository.rs:
```diff
@@ -0,0 +1,208 @@
+
+#[derive(Debug, Clone, sqlx::FromRow)]
+#[allow(dead_code)]
+pub struct SmsVerification {
```
Can we rename this to make it clearer that it is a db entity? `SmsVerificationEntity`?

SeverinAlexB commented on src/sms_verification/repository.rs:
```diff
@@ -0,0 +1,208 @@
+#[derive(Debug, Clone, sqlx::FromRow)]
+#[allow(dead_code)]
+pub struct SmsVerification {
+    pub unique_id: i32,
```
Just id assuming this is the database id?
```suggestion
    pub id: i32,
```

SeverinAlexB commented on src/sms_verification/repository.rs:
```diff
@@ -0,0 +1,208 @@
+                    .to_owned(),
+            )
+            .map_err(|e| DbError::QueryBuild(format!("Failed to build insert query: {}", e)))?
+            .to_owned();
```
This subquery is hard for me to grasp. Can you elaborate it does?

86667 replied:
Added a little extra to the comments, is that more clear?

SeverinAlexB commented on src/sms_verification/repository.rs:
```diff
@@ -0,0 +1,208 @@
+    }
+
+    /// Check if an active (pending) verification session exists for a phone number.
+    pub async fn check_pending_exists(
```
Hmm, can we get a better name here? "Check" doesn't convey in which case this method errs.

```suggestion
    pub async fn err_if_no_active_verification()
```

SeverinAlexB commented on README.md:
````diff
@@ -1,3 +1,27 @@
+
+```bash
+# Run all tests
+TEST_PUBKY_CONNECTION_STRING=postgres://postgres:postgres@localhost:5432/pubky_homegate?pubky-test=true cargo test
````
Wrong env variable. See .env.test

SeverinAlexB commented on src/e2e/wiremock_helpers.rs:
```diff
@@ -0,0 +1,93 @@
+            homeserver_pubky: "test-homeserver-pubky".to_string(),
+            max_verified_sessions: 10,
+        }
+    }
```
This shouldnt be here. The full env config for homegate has nothing to do with the mock servers. Please structure better.

86667 committed:
test: a few more tests

SeverinAlexB commented on src/sms_verification/tests.rs:
```diff
@@ -0,0 +1,1408 @@
+use sqlx::PgPool;
+use std::net::IpAddr;
+
+#[sqlx::test]
+async fn test_service_full_verification_flow(pool: PgPool) {
+    let servers = WiremockServers::start().await;
+    let service = create_service_with_mocked_apis(pool.clone(), &servers).await;
+    let db = SqlDb::test(pool.clone()).await;
```
Nice tests. Can any of these e2e tests be moved over to the corresponding modules. For example, this `sms_verification_service.rs` clearly belongs to the sms verification module. Therefore it should also live there.

SeverinAlexB replied:
> Also, ready for review again. I put all the tests in a single place so that wiremock is contained in /e2e, lets keep it like that for now.

Wiremocks should be split too and moved to the according module folder. Happy to hop on a call if you have questions.

SeverinAlexB commented on src/e2e/http.rs:
```diff
@@ -0,0 +1,304 @@
+// Full-Flow HTTP Integration Tests
+
+#[sqlx::test]
+async fn test_http_full_verification_flow(pool: PgPool) {
```
This seems to test the sms verification flow. This should live in the sms_verification module if I am not missing anything.

86667 replied:
The idea is that this tests the http return types for happy path, whereas sms_verification tests are at module level. I may reconsider the testing stratergy - we could get away with doing all testing e2e for this codebase. It was more of an experiment, i think its fine for now.

SeverinAlexB replied:
https://github.com/pubky/homegate/pull/2#discussion_r2605747755

SeverinAlexB replied:
Things that belong together should stick together.

SeverinAlexB commented on src/infrastructure/sql/error.rs:
```diff
@@ -0,0 +1,10 @@
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum DbError {
+    #[error("Database error: {0}")]
+    Database(#[from] sqlx::Error),
+
+    #[error("Query building error: {0}")]
+    QueryBuild(String),
+}
```
To make your life easy: Use `expect()` on query builds.
If you write a incorrect Select statement, you should be able to catch this relatively quickly for example in a test case.

Only changing the values usually doesn't lead to a query build error.

This way, you can always use the `sqlx::Error` error.

86667 replied:
ty :+1:

SeverinAlexB commented on src/infrastructure/database/migrations/m20251201_create_sms_verifications.rs:
```diff
@@ -0,0 +1,81 @@
+            .col(ColumnDef::new("prelude_id").text().not_null())
+            .col(
+                ColumnDef::new("created_at")
+                    .timestamp_with_time_zone()
```
created_at should not have a timezone. Same for finalized at

SeverinAlexB commented on src/infrastructure/database/migrations/m20251201_create_sms_verifications.rs:
```diff
@@ -0,0 +1,81 @@
+                    .default(sea_query::Expr::current_timestamp()),
+            )
+            .col(ColumnDef::new("finalised_at").timestamp_with_time_zone())
+            .col(ColumnDef::new("signup_code").binary())
```
NIT: Why turn something like `TBNM-K9EAC-PPE6` into binary and then later back? Why not use text directly?

86667 replied:
Good catch - This was AI code which i didnt check closely enough.

SeverinAlexB commented on src/sms_verification/repository.rs:
```diff
@@ -0,0 +1,208 @@
+            .columns(["phone_number", "prelude_id"])
+            .select_from(
+                Query::select()
+                    .expr(Expr::value(phone_number))
```
Please hash the phone number

SeverinAlexB commented on src/sms_verification/repository.rs:
```diff
@@ -0,0 +1,208 @@
+impl SmsVerificationRepository {
+    pub fn new(db: SqlDb) -> Self {
+        Self { db }
+    }
```
NIT: Using the SqlDb directly here prevents the usage of transactions. Not sure if this is needed in you code structure but the `UnifiedExecutor` that existed initially in this code base would have made using either the db directly or transactions very easy.

86667 replied:
I removed since I dont see a need for db transactions so we can save ourselves the (slight) complexity. Can always add it back if the need arises

SeverinAlexB reviewed (changes requested):
Structure is pretty good now! The executor looks flawless. The routers are properly split up. 
This time, I went deeper into the code details. I didn't find any major flaws. The overall code should work as is. I added comments to various details that would improve the code quality though.

86667 committed:
fix: Some misc code quality improvements

86667 committed:
fix:
	- Pattern for having Request params validation performed as a part of Axum's deserialisation
	- More code cleanup

86667 committed:
feat: Phone number hashing

SeverinAlexB commented on src/sms_verification/hasher_argon2id.rs:
```diff
@@ -0,0 +1,121 @@
+
+impl PhoneHasher {
+    /// Creates a new PhoneHasher with the given pepper.
+    pub fn new(pepper: String) -> Self {
```
As far as I know, salts should be unique for each hash, not a global value for all hashes.

https://grok.com/share/bGVnYWN5LWNvcHk_98b3b62e-5a47-4595-adf9-eb677485873c

86667 replied:
This is whats known as a pepper, not a salt. Its a single secret value used as salt for all hashes. 
Salts are not suitable here because we need to be able to index over phone_number

SeverinAlexB commented on src/sms_verification/hasher_argon2id.rs:
```diff
@@ -0,0 +1,121 @@
+    MissingHashOutput,
+}
+
+/// Hashes phone numbers using Argon2id with a global pepper for rainbow table resistance.
```
Cool hash algo choice.

Please elaborate on the hash properties in the description. The configured argon2id params will use a substantial amount of RAM + use a lot of time to compute the hash (380ms per hash?). Devs should be aware of the implications + our devops needs to know how to scale the server appropriately.

SeverinAlexB commented on src/sms_verification/hasher_argon2id.rs:
```diff
@@ -0,0 +1,121 @@
+
+    #[error("Hash output is missing from Argon2 result")]
+    MissingHashOutput,
+}
```
At least the `SaltEncodingError` is unnecessary. Not sure if the others are needed. Why should a hash fail?

86667 replied:
Why is it unnecessary? Better to be defensive, no?

SeverinAlexB replied:
`encode_b64` function description: `Returns Error if the slice is too long.`. At the same time, you restrict the input to 16 bytes. Therefore, the error should never occur. Without having researched when the other errors occur, is it possible to never make them occur?

86667 replied:
Removed all errors :+1:

SeverinAlexB commented on src/sms_verification/hasher_argon2id.rs:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    /// Hashes a phone number using Argon2id with a deterministic salt.
+    /// Returns an error if hashing fails (extremely rare)
```
`if hashing fails (extremely rare)`?

86667 replied:
removed

SeverinAlexB commented on src/sms_verification/hasher_argon2id.rs:
```diff
@@ -0,0 +1,121 @@
+use argon2::{
```
This module feels AI generated, has lots of small things in it that need to be reviewed. Please review AI generated code more carefully. Correct me if I am wrong.

86667 replied:
Not wrong, cleanup mainly from your suggestions. Also removed pointless tests.

SeverinAlexB commented on src/infrastructure/sql/error.rs:
```diff
@@ -0,0 +1,13 @@
+#[derive(Error, Debug)]
+pub enum DbError {
+    #[error("Database error: {0}")]
+    Database(#[from] sqlx::Error),
+
+    #[error("Database item not found: {0}")]
+    NotFound(String),
+
+    #[error("Phone number hashing error: {0}")]
+    HashingError(#[from] crate::sms_verification::PhoneHasherError),
+}
```
A hashing error is not a db error

86667 replied:
The hasher is called by sms_verifiaciton's repository code, which returns `DbError`. This does feel like the correct place for the hashing (only repository cares about storage format) and the correct error type to be returned from this code. 

Do you think it would be better to hash at the service layer?

SeverinAlexB replied:
> only repository cares about storage format

Isn't not really a storage format because it is not reversible. Any code that reads rows in the table needs to contend itself with the hash.

At the moment, you are lucky that `SmsVerificationEntity` is not leaking outside of the repo. Usually, the entity always leaks. In the future, it might very well leak.

> Do you think it would be better to hash at the service layer?

Potentially yes but I haven't thought about this a lot.

At the same time, my intuition tells me that the hash never should fail at all.

Building a proper error struct is not easy. I don't have the perfect solution here. Continue however you like.

86667 replied:
Alright, removed hasher errors to remove the need for this discussion

SeverinAlexB commented on src/infrastructure/database/migrations/m20251201_create_sms_verifications.rs:
```diff
@@ -0,0 +1,81 @@
+                    .auto_increment()
+                    .primary_key(),
+            )
+            .col(ColumnDef::new("phone_number").text().not_null())
```
Should be named `phone_number_hash` to avoid confusion?

SeverinAlexB commented on src/sms_verification/http.rs:
```diff
@@ -0,0 +1,116 @@
+    config: &EnvConfig,
+    db: crate::infrastructure::database::SqlDb,
+) -> Result<Router, HttpServerError> {
+    let state = AppState::new(config, db).map_err(|_| HttpServerError::RouterInit)?;
```
AppState never throws an error so this entire RouterInit error is redundant.

SeverinAlexB commented on src/sms_verification/http.rs:
```diff
@@ -0,0 +1,116 @@
+                tracing::error!(error = %err, "Failed to communicate with SMS provider");
+                (
+                    StatusCode::BAD_GATEWAY,
+                    "external_service_error",
+                    "Failed to communicate with external API".to_string(),
+                )
+            }
```
Probably a Internal Server Error too as Bad Gateway is only used in proxies as far as I know, never services.

SeverinAlexB commented on src/sms_verification/phone_number.rs:
```diff
@@ -0,0 +1,99 @@
+/// A validated phone number in E.164 format.
+/// E.164 format: starts with +, followed by 1-15 digits (e.g., +30123456789)
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PhoneNumber(String);
```
Nice

SeverinAlexB commented on src/sms_verification/repository.rs:
```diff
@@ -0,0 +1,247 @@
+#[derive(Debug, Clone, sqlx::FromRow)]
+pub struct SmsVerificationEntity {
+    pub id: i32,
+    pub phone_number: String,
```
Please be careful with changes. This is now a phone_number_hash, not a phone_number anymore

SeverinAlexB commented on src/sms_verification/error.rs:
```diff
@@ -0,0 +1,23 @@
+    TooManyVerifiedSessions,
+
+    #[error("No active verification session for phone number: {0}")]
+    NoActiveVerification(String),
```
String again. Could be of type PhoneNumber. Same goes for `InvalidPhoneNumber`. Please update all string phonenumber string types to the struct. Consistency is key here.

86667 replied:
InvalidPhoneNumber is correct as its the str value which failed to instantiate a `PhoneNumber`

SeverinAlexB commented on src/infrastructure/http/extractors.rs:
```diff
@@ -0,0 +1,223 @@
+}
+
+impl<T, Raw, E, S> FromRequest<S> for ValidatedJson<T, Raw, E>
+where
```
I have not tried it myself but I am pretty sure there is a better way to handle this error aka without having a "raw" request struct for each request type.

Some grok (see my second prompt) https://grok.com/share/bGVnYWN5LWNvcHk_40dec5cc-5b6e-4eb8-9ac3-cc33c97e77e3

Again, all unverified so I might be wrong.

86667 replied:
I dont believe so. We're trying to achieve custom type deserialisation with custom validation. We need to at some point tell the deserialiser to parse the request param into a String. The Validator crate helps for basic validation checks, but cannot spit out the custom type, it still performs its validation on String.

I wonder if the correct DDD way to solve this is to have API-level structs on which basic validation can be performed with primitive types, then module-level structs which perform further validation and express custom types. 

We are attempting to merge the responsibilities of two distinct domains here?

86667 replied:
That would allow us to reference those API-level structs in the API documentation too. For a client using the API seeing
```
RequestType {
    // Phone number of format.. eg +3764987
    phone_number: String
}
```
Is easier to understand than:
```
RequestType {
   // Phone number string of format.. eg "+3764987"
    phone_number: PhoneNumber
}
```

SeverinAlexB replied:
> but cannot spit out the custom type,

serde::deserialize which is used in the Json extractor automatically should turn it into a PhoneNumber. No in-between step necessary.

See this vibe coded example I made. It includes custom types and tests that showcase the flow.
https://github.com/pubky/homegate/blob/json-extractor/src/infrastructure/http/json_extractor.rs

SeverinAlexB reviewed (changes requested):
Getting better and better

86667 committed:
feat: Generalise hasher and improve code a little

86667 committed:
fix: Split tests into modules, few other fixes

86667 commented:
@SeverinAlexB  Thanks!

- Generalised hasher 
- Cleaned up all small suggestions
- Split up wiremock tests

86667 committed:
fix: Remove hasher errors

SeverinAlexB commented on src/infrastructure/config.rs:
```diff
@@ -0,0 +1,104 @@
+    pub homeserver_pubky: String,
+    #[serde(default = "default_max_verified_sessions")]
+    pub max_verified_sessions: u32,
+    pub phone_number_pepper: String,
```
Instead of configuring the pepper (which users will fail to do or fail to find an actual random value and not backup this value), can we create a new db table `Settings` and add the pepper there?

If not, we need a really good description in .env.example on how to generate this properly.

86667 replied:
That would defeat the purpose of the pepper! Its the added entropy to help blind the phone numbers in the db. I'll significantly improve the docs for it

SeverinAlexB reviewed (changes requested):
Getting very close to being done

SeverinAlexB commented on src/infrastructure/config.rs:
```diff
@@ -0,0 +1,104 @@
+pub struct EnvConfig {
+    #[serde(default)]
+    pub database_url: ConnectionString,
+    pub http_listen_socket: SocketAddr,
```
Can we get a default socket here?

SeverinAlexB commented on src/infrastructure/http/server.rs:
```diff
@@ -3,45 +3,55 @@ use std::{
+    pub fn create_router(sms_verification_router: Router) -> Router {
+        Router::new()
+            .route("/", get(root))
+            .nest("/v1/sms_verification", sms_verification_router)
```
I feel v1 is unnecessary. Can we remove it?

86667 committed:
fix: Pepper to be stored on disk. Set default http_listen_socket

BitcoinErrorLog commented:
I think we should avoid the IP thing and go with phone limiting. We dont need limiting for LN, because the limiting is the price, and we would run into similar issues to IP limiting when we limit a hub node.

SeverinAlexB commented on src/sms_verification/http.rs:
```diff
@@ -0,0 +1,116 @@
+        .send_code(request.into_inner())
+        .await?;
+    Ok(Json(response))
+}
```
I am getting a `HTTP 500 Internal Server Error` when I send an invalid code.

<img width="1162" height="1194" alt="CleanShot 2025-12-11 at 09 11 01@2x" src="https://github.com/user-attachments/assets/f909a385-72ae-408d-bcec-bb008c575a04" />

SeverinAlexB commented on src/sms_verification/http.rs:
```diff
@@ -0,0 +1,116 @@
+    State(state): State<AppState>,
+    RequestOrigin(ip_address): RequestOrigin,
+    request: ValidatedJson<
+        CreateVerificationRequest,
+        CreateVerificationRequestRaw,
+        SmsVerificationError,
+    >,
+) -> Result<Json<CreateVerificationResponse>, SmsVerificationError> {
+    let response = state
+        .sms_verification
+        .create_verification(request.into_inner(), ip_address)
+        .await?;
+    Ok(Json(response))
+}
```
When trying to create a second phone verification with the same phone number. I get a `HTTP 500 Internal Server Error`. I should be able to resend an sms code in case I didn't receive it.

<img width="1116" height="1258" alt="CleanShot 2025-12-11 at 09 16 42@2x" src="https://github.com/user-attachments/assets/212e9644-36de-4056-a125-1e9dccf98070" />

86667 replied:
This looks like a 429 from Prelude. I'll handle it by mapping to a 429

SeverinAlexB commented on src/sms_verification/http.rs:
```diff
@@ -0,0 +1,116 @@
+        .send_code(request.into_inner())
+        .await?;
+    Ok(Json(response))
+}
```
No idea how I am getting this DB error now. 

```
2025-12-11T09:18:13.157944Z ERROR request{method=POST uri=/sms_verification/verify_code version=HTTP/1.1}: homegate::sms_verification::http: Database operation failed error=Database item not found: vrf_00000000000000000000000000
```

<img width="1218" height="1268" alt="CleanShot 2025-12-11 at 10 20 50@2x" src="https://github.com/user-attachments/assets/08b8e06b-a698-4a94-91c9-a696853ffeaf" />

86667 replied:
Handled

86667 committed:
feat: weekly and yearly verification limits

86667 committed:
fix: handle prelude apis 429 error and validate code value ourselves

86667 committed:
feat: Simplify types and http responses

86667 committed:
fix: marking verifications as FAILED

SeverinAlexB commented on src/sms_verification/error.rs:
```diff
@@ -0,0 +1,51 @@
+    #[error(
+        "Invalid phone number format: {0}. Phone number must be in E.164 format (e.g., +30123456789)"
+    )]
+    InvalidPhoneNumber(String),
```
NIT: This error is not used

SeverinAlexB commented on src/sms_verification/error.rs:
```diff
@@ -0,0 +1,51 @@
+    Blocked,
+
+    #[error("Invalid code format: {0}. Code must be exactly 6 digits (0-9)")]
+    InvalidCode(String),
```
NIT: I think this error is not used.

SeverinAlexB commented on src/lib.rs:
```diff
@@ -1,6 +1,12 @@
+
+pub use shared::HomeserverAdminAPI;
+
+pub use sms_verification::{SmsVerificationError, SmsVerificationService, ValidateCodeResponse};
```
NIT: This is never used as a library as far as I know. Remove `lib.rs`?

SeverinAlexB commented on src/sms_verification/service.rs:
```diff
@@ -0,0 +1,184 @@
+        let annual_count = self
+            .repository
+            .count_verified_sessions_in_last_days(phone_number, 365)
+            .await?;
```
Calling `send_code` takes a whooping 1.65s on my Macbook M1. I assume this is likely even slower on a Google server. Probably caused by the fact that the endpoint hashes the phone number 3 times. Very slow.

86667 replied:
:laughing: - fixed

SeverinAlexB commented on src/sms_verification/http.rs:
```diff
@@ -0,0 +1,93 @@
+) -> Result<Json<ValidateCodeResponse>, SmsVerificationError> {
+    let response = state.sms_verification.validate_code(request).await?;
+    Ok(Json(response))
+}
```
Please check your logic


<img width="1598" height="1162" alt="CleanShot 2025-12-15 at 09 15 40@2x" src="https://github.com/user-attachments/assets/3bdad6e0-dca6-40ab-aff4-1c3705014253" />

SeverinAlexB commented on src/sms_verification/service.rs:
```diff
@@ -0,0 +1,184 @@
+                id
+            );
+            return Err(SmsVerificationError::Blocked);
+        }
```
NIT: Should we even create a db record when the phone number is blocked? What's the purpose?

86667 replied:
The thinking is that its easier to just store everything for debugging/audit/whatever as the records are small in size and quantity. Easier than having to remember what is stored and what is not

86667 replied:
Ive removed that double `if let` too

SeverinAlexB commented on src/sms_verification/service.rs:
```diff
@@ -0,0 +1,184 @@
+        Ok(())
+    }
+
+    /// Initiates a phone number verification process
+    pub async fn create_verification(
+        &self,
+        request: CreateVerificationRequest,
+        ip_address: IpAddr,
+    ) -> Result<(), SmsVerificationError> {
```
NIT: Race condition possible?

The user only has 1 verification left after he reaches the limit. Two requests can be made at almost at the same time to surpass the limit.

First request:
1. Check verification limits -> Pass

Second Request, after the first request "Check verification limits" but before the "Create Verification":
1. Check verification limits -> Pass

This way, he can surpass our limit.

NIT because it's not critical. Could be fixed by using transactions.

86667 replied:
There can only be one verification window per phone number, so the second would be a `Retry`

86667 replied:
Actually, that may be true but this is still flaky. I'll use txs

SeverinAlexB commented on src/e2e/http.rs:
```diff
@@ -0,0 +1,319 @@
+        body_text.contains("E.164"),
+        "Error should reference E.164 format"
+    );
+}
```
NIT: A lot of these "e2e" tests actually just test the parts of the sms_verification module and are not actual e2e tests. So they should be moved to the sms_verification folder.

SeverinAlexB commented on src/e2e/http.rs:
```diff
@@ -0,0 +1,319 @@
+    response.assert_status_ok();
+
+    // Wiremock will verify the IP was sent in the request body
+}
```
This either tests
- the ip extractor itself
- or the send_code_handler

Either way, the test should be moved closed to what it tests. Either the extractor or the send_code_handler.

e2e folder is not the right place.

SeverinAlexB reviewed (changes requested):
As I want to get this PR merged, I marked most comments as NITs so feel free to ignore them. Please have a look at the non-NITs though as they need fixing to prevent problems in production.

86667 committed:
fix: rm lib, hash once per request

86667 committed:
feat: Store db in Appstate, pass through to service entry points, mv http tests to module and rm dup ip extraction tests

86667 committed:
fix: handle country blocked error and remove phone number from error msgs

86667 committed:
fix: supersede verificaiotn with different prelude_id

SeverinAlexB reviewed (approved):
Let's go!
